unix: add UV_HANDLE_IPV6 flag to tcp and udp handles

saghul opened the pull request:
Aligns behavior with Windows. Currently the flag is used to pick the default multicast interface when resetting it.

/cc @indutny, @bnoordhuis

saghul committed:
unix: add UV_HANDLE_IPV6 flag to tcp and udp handles

Nodejs-Jenkins commented:
Thank you for contributing this pull request! Here are a few pointers to make sure your submission will be considered for inclusion.

Commit saghul/libuv@24bcdf60bb182755108300420f0a06cb6225dd2c has the following error(s):
- First line of commit message must be no longer than 50 characters

The following commiters were not found in the CLA:
- Saúl Ibarra Corretgé

You can fix all these things _without_ opening another issue.

Please see [CONTRIBUTING.md](https://github.com/joyent/node/blob/master/CONTRIBUTING.md) for more information

indutny commented on src/unix/tcp.c:
```diff
@@ -89,8 +89,11 @@ int uv__tcp_bind(uv_tcp_t* tcp,
   errno = 0;
   if (bind(tcp->io_watcher.fd, addr, addrlen) && errno != EADDRINUSE)
     return -errno;
-
```
No need to change whitespace.

saghul replied:
I thought I'd group the errono handlign statements there and add my condition afterwards, but I can revert that part.

indutny replied:
oh, ok then.

indutny commented on src/unix/udp.c:
```diff
@@ -619,8 +622,13 @@ int uv_udp_set_multicast_interface(uv_udp_t* handle, const char* interface_addr)
     memset(&addr_st, 0, sizeof addr_st);
-    addr_st.ss_family = AF_INET;
-    addr4->sin_addr.s_addr = htonl(INADDR_ANY);
+    if (!(handle->flags & UV_HANDLE_IPV6)) {
```
May be inverse the condition? I doubt that `!` is really needed here.

indutny replied:
Also seems like a behavior change, does it need any test?

saghul replied:
I'd rather write it like this because we always do IPv4 first and IPv6 after. There is no behavior "change" as setting the default interface for IPv6 never worked before this patch :-)

indutny replied:
Could you add a test for default interface then? ;)

saghul replied:
Won't work on OSX, unfortunately, but I can write one.

indutny replied:
You could ifdef it if needed, that would be fine.

bnoordhuis replied:
> I'd rather write it like this because we always do IPv4 first and IPv6 after.

The counter-argument is that negation requires more mental effort to parse.  Just a tiny bit more, of course, but tiny bits add up.

indutny commented:
Looks good, see comments.

bnoordhuis commented on src/unix/tcp.c:
```diff
@@ -89,8 +89,11 @@ int uv__tcp_bind(uv_tcp_t* tcp,
   tcp->delayed_error = -errno;
+
+  if (addr->sa_family == AF_INET6)
+    tcp->flags |= UV_HANDLE_IPV6;
```
Here's an edge case to consider: what if I bind the the socket to a IPv4 address first, then connect to a IPv6 address (or vice versa)?  Should UV_HANDLE_IPV6 be set or not?

saghul replied:
I'd say no. The meaning the flag currently has is basically "did we create the socket with AF_INET6 or not?". Right now the socket is always created right before binding, should that change in the future, we might need to revisit.

saghul commented:
I rewrote the flag test (looked better once I wrote it that way :-)) and updated the multicast test to use NULL, that is, the default interface. (except on OSX, because it doesn't work)

saghul commented:
No (new) failing tests :-)

saghul commented:
Any further comments?

bnoordhuis commented:
None from me.

saghul commented:
Kewl, thanks for confirming, Ben!

saghul commented:
Landed in https://github.com/joyent/libuv/compare/3558d65d2f17...7a4c42a0548b
